### DISCUSSION TITLE
Why qps does not  go up when I use more client concurrently?

### DISCUSSION THREAD

**Discussion by xiaojuan8** [milvus-io/milvus#22066] [Q&A and General discussion] [answered]
Hi, I have 8 replicas of querynode. And I want to use pressure-test for searching results. I start 100 threads for concurrently request querynodes for searching. When I add thread num to 500 threads at 17:07 , searching qps shows no changes on grafana panel, and I find that searching latency does not show any differences：
![image](https://user-images.githubusercontent.com/74544828/217486712-7921f8a5-acd0-492e-8582-ce94cb0500f4.png)

CPU is not fully used：
![image](https://user-images.githubusercontent.com/74544828/217487039-76b42f51-8405-4867-8896-132b505e3be0.png)

**Comment by xiaojuan8** [comment]
my python client script：
`from pymilvus import connections
from pymilvus import CollectionSchema, FieldSchema, DataType, Collection
import h5py,random
from random import randint
import numpy as np
from threading import Thread, Barrier
from Counter import MyCounter
import time

THREAD_NUM = 100 # control concurrency
nq = 10
filename = "sift-128-euclidean.hdf5"
f = h5py.File(filename, 'r')
test_set = []
for key in f.keys():
  if key == 'test':
    test_set = list(f[key][:])

connections.connect(alias="default", host='10.185.14.191', port='32167')

print("connection finished")

collection_name = "SIFT"

collection = Collection(
name=collection_name, 
)

collection.load()

print("collection load done")

search_params = {"metric_type": "L2", "params": {"nprobe": 10}, "offset": 5}

size = len(test_set)
def search(counter : MyCounter):
  counter.inc()
  for j in range(10000):
    query = []
    for i in range(nq):
      index = randint(1, size - 1)
      query.append(test_set[index])
    results = collection.search(
      data=query, 
      anns_field="emb", 
      param=search_params,
      limit=10, 
      expr=None,
      consistency_level="Eventually"
    )
    print("query search done")
    print("query results : ")
    print(results)
  counter.dec()

count = MyCounter()
for i in range(THREAD_NUM):
  t = Thread(target=search, args=(count, ))
  t.start()

while count.value() > 0:
  print("waiting...")
  time.sleep(1)

collection.release()`

**Comment by yhmo** [comment, marked as answer]
There is a global lock in python thread model: https://realpython.com/python-gil/
"This means that only one thread can be in a state of execution at any point in time. The impact of the GIL isn’t visible to developers who execute single-threaded programs, but it can be a performance bottleneck in CPU-bound and multi-threaded code."

You can try multi-process instead of multi-thread.

**Reply by xiaojuan8** [reply]
Thanks, it works.
